datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+        subdatasets_only=Parameter(
+            args=("-s", "--subdatasets-only"),
+            action="store_true",
+            doc="""whether to exclude top level dataset."""),
+        passthrough=Parameter(  # TODO  could be of use for `run` as well
+            args=("-p", "--passthrough"),
+            action="store_true",
+            doc="""For command line commands, pass-through their output to 
+            the screen instead of capturing/returning as part of the result records."""),
+        jobs=jobs_opt,
+        # TODO: might want explicit option to either worry about 'safe_to_consume' setting for parallel
+        # For now - always safe
+    )
+
+    @staticmethod
+    @datasetmethod(name='foreach')
+    @eval_results
+    def __call__(
+            cmd=None,
+            python=False,
+            dataset=None,
+            fulfilled=None,
+            recursive=False,
+            recursion_limit=None,
+            # contains=None,
+            bottomup=False,
+            subdatasets_only=False,
+            passthrough=False,
+            jobs=None
+            ):
+        if not cmd:
+            lgr.warning("No command given")
+            return
+        if python and len(cmd) > 1:
+            # yoh decided to avoid unnecessary complication/inhomogeneity with support
+            # of multiple Python commands for now
+            raise ValueError(f"Please provide a single Python expression. Got {len(cmd)}: {cmd!r}")
+        refds = require_dataset(
+            dataset, check_installed=True, purpose='foreach execution')
+        pwd = getpwd()  # Note: 'run' has some more elaborate logic for this
+        subdatasets_it = refds.subdatasets(
+            fulfilled=fulfilled, recursive=recursive, recursion_limit=recursion_limit,
+            bottomup=bottomup,
+            result_xfm='paths'
+        )
+
+        if subdatasets_only:
+            datasets_it = subdatasets_it
+        else:
+            if bottomup:
+                datasets_it = chain(subdatasets_it, [refds.path])
+            else:
+                datasets_it = chain([refds.path], subdatasets_it)
+
+        if not python:
+            protocol = NoCapture if passthrough else StdOutErrCapture
+
+        def run_cmd(dspath):
+            ds = Dataset(dspath)
+            status_rec = get_status_dict(
+                'foreach',
+                ds=ds,
+                path=ds.path,
+                command=cmd
+            )
+            if not ds.is_installed():
+                yield dict(
+                    status_rec,
+                    status="impossible",
+                    message="not installed"
+                )
+                return
+            # For consistent environment (Python) and formatting (command) similar to `run` one
+            # But for Python command we provide actual ds and refds not paths
+            placeholders = dict(
+                pwd=pwd,
+                ds=ds if python else ds.path,
```

Comment:
For consistency with `run`, perhaps it's worth adding `ds_path` too?

Reply:
so in Python case it would differ (`ds` would be instance and `ds_path` a string) while in CMD they would be identical, right?

Reply:
Yes, that was my thinking (though maybe that'd be confusing?)

Reply:
stroke me (didn't test/check): can't we have `{ds}` placeholder with `{ds.path}` etc (by `.format` accessing that `ds` attributes) in `run` etc?  then if we add `{ds}` (and `{refds}`) to `run` I would just may be drop `ds_path` here completely so people could just use `{ds.path}` if so desired regardless if it is CLI command or python eval.

Reply:
note: I used the wrong name above.  It's `dspath` for `run`.  

We could expose add `ds=` to `run`, and use `ds.path`.  I don't think we should, though, unless there's some other attribute of `ds` that we're actually interested in exposing.

Reply:
I do not have an immediate pragmatic use case for `run` as well ATM, but I can see how the ability to access `{ds.pathobj}` for cross platform paths specifications etc or `{ds.config}` for configuration variables (which `git config` would not be sufficient due to not be aware of all additional locations) would be of benefit at least for `foreach`.  
I just do not see why then `run` should not be allowed to do the same (i.e., any possible side-effects?) and thus reduce user cognitive load in remembering which placeholders are available in `run/rerun` and which in `foreach`.

Reply:
If there's not an immediate use case, I'd rather not add it.  When you're dealing with `run` commands, you're talking about supporting things indefinitely.  I don't think it's a good place to add things just because.

Reply:
ok, to reduce user mental load I think I will add `{dspath}` as `= {ds.path}`  for `foreach`, even though as discussed/agreed exposing `{ds}` to `run` isn't .